pkg/csi_driver/multishare_controller.go

```diff
@@ -87,6 +88,33 @@ func (m *MultishareController) CreateVolume(ctx context.Context, req *csi.Create
 		return nil, err
 	}
 
+	param := req.GetParameters()
+	// If we are creating a new instance, we need pick an unused CIDR range from reserved-ipv4-cidr
+	// If the param was not provided, we default reservedIPRange to "" and cloud provider takes care of the allocation
+	if instance.Network.ConnectMode == privateServiceAccess {
+		if reservedIPRange, ok := param[paramReservedIPRange]; ok {
+			instance.Network.ReservedIpRange = reservedIPRange
+		}
+	} else if reservedIPV4CIDR, ok := param[paramReservedIPV4CIDR]; ok {
+		reservedIPRange, err := m.controllerServer.reserveIPRange(ctx, &file.ServiceInstance{
+			Project:  instance.Project,
+			Name:     instance.Name,
+			Location: instance.Location,
+			Tier:     instance.Tier,
+		}, reservedIPV4CIDR)
+
+		// Possible cases are 1) CreateInstanceAborted, 2)CreateInstance running in background
+		// The ListInstances response will contain the reservedIPRange if the operation was started
+		// In case of abort, the CIDR IP is released and available for reservation
+		defer m.controllerServer.config.ipAllocator.ReleaseIPRange(reservedIPRange)
```

Comment:
note this defer is executed at the end of the function call. this means lock will be released only after instance + share create complete in worst case.  we can technically release lock at the end of instance create. However since share create is relatively much faster, this should be fine for now. something to keep in mind.